### PR TITLE
Feat: ack dispatching to contracts

### PR DESCRIPTION
**Pull request by pr0n00gler**
This PR solves:
* https://p2pvalidator.atlassian.net/jira/software/projects/LSC/boards/121?selectedIssue=LSC-68

What's changed:
* now all IBC acknowledgements are dispatched to appropriate smart contracts which submitted an interchain transactions;
* msgSubmit.Packet.SourcePort contains address of the contract to dispatch acknowledgement later and an .Owner which allows the contract to have multiple interchain accounts;

**IMPORTANT NOTICE**: Since we don't have contracts yet which can send interchain transactions, we allow anyone to set "callback ack contract", because we want sudo-demo to work. Because of that some tests are not passed (but if you enable production ready code in `x/interchaintxs/types.go`, tests will pass). It's either demo or tests for now

### CODE REVIEW

File: x/interchainqueries/keeper/keeper_test.go [hunk 1]
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/lidofinance/gaia-wasm-zone/x/interchainqueries/keeper"
 	iqtypes "github.com/lidofinance/gaia-wasm-zone/x/interchainqueries/types"
 	itypes "github.com/lidofinance/gaia-wasm-zone/x/interchainqueries/types"
```
Comment (sotnikov-s) [inline x/interchainqueries/keeper/keeper_test.go]: lol the same import with different names:
```go
        iqtypes "github.com/lidofinance/gaia-wasm-zone/x/interchainqueries/types"
	itypes "github.com/lidofinance/gaia-wasm-zone/x/interchainqueries/types"
```
please fix. would be nice and clean to have it similar to the import you've introduced a line below

File: x/interchaintxs/types/types.go [hunk 2]
```diff
@@ -1 +1,60 @@
...
+	return ICAOwner{contractAddress: sdkContractAddress}, nil
+}
+
+func ICAOwnerFromPort(port string) (ICAOwner, error) {
+	splittedOwner := strings.Split(strings.ReplaceAll(port, icatypes.PortPrefix, ""), delimiter)
+	if len(splittedOwner) < 2 {
+		return ICAOwner{}, sdkerrors.Wrap(ErrInvalidICAOwner, "invalid ICA owner format")
+	}
+
+	contractAddress, err := sdk.AccAddressFromBech32(splittedOwner[0])
```
Comment (swelf19) [inline x/interchaintxs/types/types.go:46]: is there any restriction portid have to be = "portID + delimiter + address"? Yeah, we use address as an owner, but i do not see any reason why others have to either

Comment (mmikev) [reply x/interchaintxs/types/types.go:46]: Didn't quite get you. Could you elaborate a little bit, please?

File: x/interchaintxs/types/types_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+			desc:            "invalid",
+			contractAddress: "invalid_contract",
+			expectedErr:     types.ErrInvalidAccountAddress,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			icaOwner, err := types.NewICAOwner(tc.contractAddress, tc.owner)
+			if tc.expectedErr == nil {
+				require.NoError(t, err)
+				require.Equal(t, icaOwner.GetContract().String(), tc.contractAddress)
```
Comment (sotnikov-s) [inline x/interchaintxs/types/types_test.go]: don't you mean to also test the owner and the icaOwner.String() method here?

File: x/interchaintxs/types/types.go [hunk 4]
```diff
@@ -1 +1,60 @@
...
+	// this means anyone can set contractAddress
+	sdkContractAddress, err := sdk.AccAddressFromBech32(owner)
+	if err != nil {
+		return ICAOwner{}, sdkerrors.Wrapf(ErrInvalidAccountAddress, "failed to decode address from bech32: %v", err)
+	}
+	return ICAOwner{contractAddress: sdkContractAddress}, nil
+}
+
+func ICAOwnerFromPort(port string) (ICAOwner, error) {
+	splittedOwner := strings.Split(strings.ReplaceAll(port, icatypes.PortPrefix, ""), delimiter)
```
Comment (sotnikov-s) [inline x/interchaintxs/types/types.go]: I think using the strings.TrimPrefix for cutting out the `icaconttoller-` part would be more correct — do I get it right that the `icaconttoller-` part is in the front of the incoming port?

File: x/interchaintxs/types/types.go (hunk 2 above)
Comment (swelf19) [reply x/interchaintxs/types/types.go:46]: looks like you are making an assumption owner = address, but owner could be any string, no?
```bash
neutrond tx interchaintxs register-interchain-account connection-0 anystring --node tcp://127.0.0.1:16657 --from demowallet1 --home ../data/test-1 --broadcast-mode=block --gas-prices 0.0025stake  -y --output json  --keyring-backend test --chain-id test-1

neutrond q interchaintxs interchainaccounts connection-0 anystring  --node tcp://127.0.0.1:16657  --home ../data/test-1   --output json   --chain-id test-1
{"interchain_account_address":"neutron1ga0ulr7dqgw8w48ugpzzswa4s5snc5npwxx9f6xz4kk4092kxt7qk92mut"}

$ neutrond q ibc channel channels --node tcp://127.0.0.1:16657
channels:
- channel_id: channel-0
  connection_hops:
  - connection-0
  counterparty:
    channel_id: channel-0
    port_id: icahost
  ordering: ORDER_ORDERED
  port_id: icacontroller-anystring
  state: STATE_OPEN
  version: '{"version":"ics27-1","controller_connection_id":"connection-0","host_connection_id":"connection-0","address":"neutron1ga0ulr7dqgw8w48ugpzzswa4s5snc5npwxx9f6xz4kk4092kxt7qk92mut","encoding":"proto3","tx_type":"sdk_multi_msg"}'
height:
  revision_height: "460"
  revision_number: "1"
pagination:
  next_key: null
  total: "0"
```

File: x/interchaintxs/keeper/ibc_handlers.go [hunk 5]
```diff
@@ -39,10 +40,10 @@ type OpenAckDetails struct {
 
 // HandleAcknowledgement passes the acknowledgement data to the Hub contract via a Sudo call.
```
Comment (sotnikov-s) [inline x/interchaintxs/keeper/ibc_handlers.go]: // HandleAcknowledgement passes the acknowledgement data to the **Hub** contract via a Sudo call.

all hub mentions should be removed

File: x/interchaintxs/types/types.go (hunk 2 above)
Comment (mmikev) [reply x/interchaintxs/types/types.go:46]: portID is (contractAddress + delimiter + owner) where the owner is any string, yes, but not contractAddress part. This allows contracts to have multiple interchain accounts on remote zone.

I understand now that `owner` is a little bit misleading name for the field, I'll rename it to `interchain_account_id` or smth

Comment (sotnikov-s) [reply x/interchaintxs/types/types.go:46]: > I understand know that owner is a little bit misleading name for the field

I was just about to ask you to do so

Comment (mmikev) [reply x/interchaintxs/types/types.go:46]: `owner` is renamed to `interchainAccountID` now

File: x/interchainqueries/keeper/keeper_test.go (hunk 1 above)
Comment (mmikev) [reply x/interchainqueries/keeper/keeper_test.go]: Fixed

File: x/interchaintxs/types/types_test.go (hunk 3 above)
Comment (mmikev) [reply x/interchaintxs/types/types_test.go]: The checks are added

File: x/interchaintxs/types/types.go (hunk 4 above)
Comment (mmikev) [reply x/interchaintxs/types/types.go]: Fixed

File: x/interchaintxs/keeper/ibc_handlers.go (hunk 5 above)
Comment (mmikev) [reply x/interchaintxs/keeper/ibc_handlers.go]: Removed

**Review by sotnikov-s — APPROVED** [review]

File: x/interchaintxs/types/types.go [hunk 6]
```diff
@@ -1 +1,60 @@
...
+	// this means anyone can set contractAddress
+	sdkContractAddress, err := sdk.AccAddressFromBech32(interchainAccountID)
+	if err != nil {
+		return ICAOwner{}, sdkerrors.Wrapf(ErrInvalidAccountAddress, "failed to decode address from bech32: %v", err)
+	}
+	return ICAOwner{contractAddress: sdkContractAddress}, nil
+}
+
+func ICAOwnerFromPort(port string) (ICAOwner, error) {
+	splittedOwner := strings.Split(strings.TrimPrefix(port, icatypes.PortPrefix), delimiter)
```
Comment (swelf19) [inline x/interchaintxs/types/types.go]: Can it be a problem, if user sends us an owner with a dot in the middle.
Original portid "icacontroller-address.ow.ner"
restored interchainAccountID from the portid - "ow" not "ow.ner"

Comment (mmikev) [reply x/interchaintxs/types/types.go]: Good point, thanks!

Fixed that and added a test for such case

**Comment by foxpy** [comment]
Is `:/start.sh` expected to fail?

I mean, it fails on my side.

**Comment by pr0n00gler** [comment]
> Is `:/start.sh` expected to fail?
> 
> I mean, it fails on my side.

With what error it fails? Because everything works for me

**Comment by foxpy** [comment]
> With what error it fails?

The first wrong thing is at `:/start.sh:9`: subcommand `$(gaia-wasm-zoned keys show test -a)` fails because a key is somehow not present in storage:

```shell
[foxpy@t14s neutron]$ gaia-wasm-zoned keys add test

- name: test
  type: local
  address: cosmos1mum7gwdsc5xtvzy7f6g52dtylswrpcrgdlhz54
  pubkey: '{"@type":"/cosmos.crypto.secp256k1.PubKey","key":"AzRGnVrb1tY0MXJ5uJV7T9KfWC5KSp28FgU2Mp9nIlwD"}'
  mnemonic: ""


**Important** write this mnemonic phrase in a safe place.
It is the only way to recover your account if you ever forget your password.

rubber skirt tent pizza unfair lounge detail draft topic cover waste perfect today ten evil please music rely nuclear door guess fit match excuse
[foxpy@t14s neutron]$ echo $?
0
[foxpy@t14s neutron]$ gaia-wasm-zoned keys list
[]

[foxpy@t14s neutron]$ 
```

To be honest I don't even have an idea of where to start troubleshooting.

**Comment by pr0n00gler** [comment]
> To be honest I don't even have an idea of where to start troubleshooting.

Just forget about it. Currently we use `make init ...` for test purposes anyway, and we have more important tasks to do than fix some unused script with some weird error.

I'll remove the script from the repo

**Comment by foxpy** [comment]
> some unused script with some weird error

I also have a feeling this error is specific for my workstation, so, yeah, let's ignore it since this script is not used anyway.

**Review by swelf19 — APPROVED** [review]

**Review by foxpy — COMMENTED** [review]
Okay, but now the error is misleading in a difference sense.

```shell
[foxpy@t14s neutron]$ gaia-wasm-zoned query interchainqueries registered-queries --node tcp://127.0.0.1:16657 | egrep '\s+id:'
  id: "1"
  id: "2"
[foxpy@t14s neutron]$ gaia-wasm-zoned query interchainqueries query-result 1 --node tcp://127.0.0.1:16657 2>&1 | grep Error
Error: rpc error: code = InvalidArgument desc = failed to get query result by query id: no query result: no query result: invalid request
[foxpy@t14s neutron]$ gaia-wasm-zoned query interchainqueries query-result 2 --node tcp://127.0.0.1:16657 2>&1 | grep Error
Error: rpc error: code = InvalidArgument desc = failed to get query result by query id: no query result: no query result: invalid request
[foxpy@t14s neutron]$ gaia-wasm-zoned query interchainqueries query-result 3 --node tcp://127.0.0.1:16657 2>&1 | grep Error
Error: rpc error: code = InvalidArgument desc = failed to get query result by query id: no query result: no query result: invalid request
[foxpy@t14s neutron]$ gaia-wasm-zoned query interchainqueries query-result 4 --node tcp://127.0.0.1:16657 2>&1 | grep Error
Error: rpc error: code = InvalidArgument desc = failed to get query result by query id: no query result: no query result: invalid request
[foxpy@t14s neutron]$ 
```

There couldn't be a query result for ids 3 and 4 since these queries don't exist in the first place.

I am not requesting changes since this issue doesn't look that bad.

**Comment by pr0n00gler** [comment]
> Okay, but now the error is misleading in a difference sense.
> There couldn't be a query result for ids 3 and 4 since these queries don't exist in the first place.

I don't think this is very important and misleading. It can be solved by adding additional code for checking existence of queries itself, but it seems a bit excessive. Moreover, the error is not wrong, there are really no results for queries 3 and 4 after all. 

It's not misleading for me at least.

**Review by foxpy — APPROVED** [review]

**Review by ratik — APPROVED** [review]

**Pull request merged**